lib/telemetry_registry.ex

```diff
@@ -0,0 +1,253 @@
+defmodule TelemetryRegistry do
+  @moduledoc """
+  TelemetryRegistry provides tools for the discovery and documentation of [telemetry](https://github.com/beam-telemetry/telemetry)
+  events within your applications.
+
+  ## Telemetry Event Definitions and Declaration
+
+  Users want to know what telemetry events are available in your library, what they mean, as well as what
+  the measurements and metadata maps contain. TelemetryRegistry creates an official standard and mechanism
+  for telemetry event declaration and optional definition. It is _highly_ encouraged to use the definition
+  format for your users.
+
+  ### Who Should Document Events?
+
+  Library authors in particular should provide event declarations for documentation and to simplify tracing.
+  Of course, everyone should document their modules!
+
+  ### Where Should I Declare Events?
+
+  Events should only be declared once, usually in the module from which it originates. Event names should _always_
+  be namespaced to your application or library. For example, if your application is an http client, your events
+  should start with the name of your application, not a generic name.
+
+  **Do:** `[:tesla, :request, :stop]`
+  **Don't:** `[:http_client, :request, :stop]`
+
+  ### Event Definition Format
+
+  Events are declared using the `telemetry_event` module attribute. The attribute accepts an event definition
+  or a plain event, though plain events are discouraged. These definitions are used for producing documentation
+  and event discovery. All definition keys are required.
+
+  ```elixir
+  %{
+    event: [:my_app, :event, :stop],
+    description: "A description of what the event is and when it is emitted",
+    measurements: "A string containing a pseudo or typespec - see examples",
+    metadata: "A string containing a pseudo or real typespec - see examples"
+  }
+  ```
+
+  ```erlang
+  #{
+    event => [my_app, event, stop],
+    description => <<"A description of what the event is and when it is emitted">>,
+    measurements => <<"A string containing a pseudo or typespec - see examples">>,
+    metadata => <<"A string containing a pseudo or real typespec - see examples">>
+  }
+  ```
+
+  #### Elixir
+  
+  Elixir does not allow for declaring a custom attribute multiple times by default. We have included macros
+  to help with this and to provide a way to include event documentation.
+
+  ```elixir
+  defmodule TestElixirApp do
+    use TelemetryRegistry
+
+    telemetry_event(%{
+      event: [:test_elixir_app, :single, :event],
+      description: "emitted when this event happens",
+      measurements: "%{duration: non_neg_integer()}",
+      metadata: "%{status: status(), name: String.t()}"
+    })
```

Comment:
Should `TelemetryRegistry` expose a formatter export for `telemetry_event` for no parens? 

```suggestion
    telemetry_event %{
      event: [:test_elixir_app, :single, :event],
      description: "emitted when this event happens",
      measurements: "%{duration: non_neg_integer()}",
      metadata: "%{status: status(), name: String.t()}"
    }
```

Reply:
:+1: